src/lsm/README.md

```diff
@@ -0,0 +1,111 @@
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
+A compaction beat is started asynchronously with `compact()` which takes a callback and
+the op that was just committed.
+
+A bar is split in half according to the "first" down beat and "middle" down beat.
+The first half of the bar compacts even levels while the latter compacts odd levels.
+Mutable table changes are sorted and compacted into the immutable table.
+The immutable table is compacted into level 0 during the odd level half of the bar.
+
+At any given point, there are at most levels/2 compactions running concurrently.
+The source level is denoted as `level_a` and the target level as `level_b`.
+The last level in the LSM tree has no target level so it is never a source level.
+
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
+* The manifest is compacted at the end of every beat.
+* The compactions' output tables are not visible until the compaction has finished.
+
+1. (first) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
+    * Start even level compactions if there's any tables to compact.
+
+2. (second) up beat of the bar:
+    * Finish ticking running even-level compactions.
+    * Assert on callback completion that all compactions are complete.
+
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
+    * Compact the immutable table if it contains any sorted values (it could be empty).
+
+4. (fourth) last beat of the bar:
+    * Finish ticking running odd-level and immutable table compactions.
+    * Assert on callback completion that all compactions are complete.
+    * Assert on callback completion that no level's visible table count overflows.
+    * Flush, clear, and sort mutable table values into immutable table for next bar.
+
+
+## Snapshots
+### Snapshots and Compaction
+
+Consider the half-bar compaction beginning at op=`X` (`12`), with `lsm_batch_multiple=M` (`8`).
+Each half-bar contains `N=M/2` (`4`) beats. The next half-bar begins at `Y=X+N` (`16`).
+
+During the compaction `X` (op=`X…Y-1`; `12…15`), each commit prefetches from the snapshot equal to
+the first op of the compaction. As shown, they continue to query the old (input) tables.
+
```

Comment:
...and this is ok because the mutable table doesn't get flushed until the end of the compact?

Reply:
Conceptually the mutable table is an "input table", to the immutable table's "output table". (But of course their compaction is just a sort+copy). This relationship is explored more in the "Snapshot Values" section.

Reply:
I guess what's missing is knowing that ops read and write to the mutable table, and that the number of writes per op is bounded so we know we can go a whole half-bar without flushing the mutable table.

Reply:
Correct, except that we go a full bar without flushing the mutable table, not a half-bar. I'll detail this in the docs.